### ISSUE TITLE
BoundingBox inaccuracy

### ISSUE THREAD

**Issue by RacoAnimech** [playcanvas/engine#1903]
### Description

```pc.BoundingBox``` stores two vectors: ```center``` and ```halfExtents```. The minimum and maximum values can be calculated from these values using ```getMin``` and ```getMax```. The problem with this is rounding errors. The currently only way to get the actual minimum and maximum is to loop over the vertexbuffer and calculate them manually.

To fix this problem ```pc.BoundingBox``` has to be remade to store the minimum and maximum values and calculate center and halfextents from them.

### Steps to Reproduce

Run this code in any project.
```js
let min = new pc.Vec3(0.3,1,1);
let max = new pc.Vec3(2.7,1,1);
let aabb = new pc.BoundingBox();
aabb.setMinMax(min, max);
console.log(aabb.getMin().x);
// output: 0.2999999999999998
console.log(aabb.getMin().x === min.x);
// output: false
```

**Comment by Maksims** [comment]
This is related to general issue with JavaScript Numbers precision. Basically what happens is this:
```javascript
0.2 + 0.4 = 0.6000000000000001
```
You can read more on this: https://stackoverflow.com/questions/11695618/dealing-with-float-precision-in-javascript

**Comment by RacoAnimech** [comment]
> This is related to general issue with JavaScript Numbers precision.  

Yes it is, but it's also preventable. Mathematically speaking storing a bounding box as center and halfextents or minimum and maximum should not matter since we can calculate them from each other. But when implemented they become distinct because of these rounding errors.

Right now ```BoundingBox``` stores center and halfextent and calculates minimum and maximum with rounding errors. An example of when these errors matter is when the ```BoundingBox``` is calculated from a vertex buffer. The minimum and maximum values might actually not be the minimum and maximum of the vertexbuffer. I suggest that ```BoundingBox``` should instead store minimum and maximum and calculate center and halfextents from these. This would solve the problem.

**Comment by slimbuck** [comment]
Yep @RacoAnimech. Presumably center & halfExtent was chosen so that _fastIntersectRay() could be fast.

I would be very surprised if the rounding on 64bits of floating point precision isn't enough though. What problems are you actually seeing with the inaccuracy?

**Comment by RacoAnimech** [comment]
I'm writing an exporter and min max values are required. It feels wrong to export incorrect values and  looping over the vertexbuffer is not an optimal solution, especially when the values are already calculated.

**Comment by slimbuck** [comment]
Yes I understand what you mean.

Just make sure it's actually a problem in your case. For example, if you're not outputting the bounds *and* meshes to binary 64bit float data, then a conversion is happening anyway, either to decimal text or 32bit floating point values. In which case the small bounding box rounding most likely makes no difference at all.

Ideally we would store min/max values as you suggest, but a change like this could have unknown knock-on effects for a relative small gain.

**Comment by mekwall** [comment]
@slimbuck We are importing glTF model and then export it back to glTF. To be able to verify that the exporter is working as expected, we expect the values in the exported glTF to be the exact same as the values in the original glTF. This rounding error causes the value to differ and it is difficult to tell between a rounding error and wrong conversion logic.

**Comment by mvaligursky** [comment]
Your code which compares floats should compare with some small epsilon to avoid this issue.
If we switched BoundingBox to store min and max, and somebody would write exporter to store center / halfExtent, they would have precision error as well. You cannot win. Design the code to handle it.

**Comment by aidinabedi** [comment]
@mvaligursky [BoundingBox already "stores" min and max](https://github.com/playcanvas/engine/blob/master/src/shape/bounding-box.js#L21). It just re-calculates them (based on center and halfExtents) on every `getMin` and `getMax`. It's quite trivial to make BoundingBox support both min/max and center/halfExtents without a big performance hit. In fact, you gain a lot of optimizations due to avoiding `getMin` and `getMax`. See my PR below.

Btw, Similar to the glTF specs both [Three.js](https://threejs.org/docs/#api/en/math/Box3) and [Babylon.js](https://doc.babylonjs.com/api/classes/babylon.boundingbox#constructor) uses min/max (instead of center/halfExtents). But again, you don't have to choose.
